icu4c/source/i18n/msgfmt.cpp

```diff
@@ -810,26 +810,31 @@ MessageFormat::getFormats(int32_t& cnt) const
     // method on this object.  We construct and resize an array
     // on demand that contains aliases to the subformats[i].format
     // pointers.
+
+    // Get total required capacity first (it's refreshed on each call).
```

Comment:
Please check the Java code: Does it have the same bug?

Reply:
I'll take a look.

Reply:
Java seems fine - they don't return the count, and allocate as appropriate.
```
public Format[] getFormats() {
     ArrayList<Format> list = new ArrayList<>();
     for (int partIndex = 0; (partIndex = nextTopLevelArgStart(partIndex)) >= 0;) {
         list.add(cachedFormatters == null ? null : cachedFormatters.get(partIndex));
     }
     return list.toArray(new Format[list.size()]);
}
```